components/motor/ulnstepper/28byj-48.go

```diff
@@ -0,0 +1,478 @@
+
+	if m.stepPosition == m.targetStepPosition {
+		err := m.enable(ctx, false)
+		return 5 * time.Millisecond, err
+	}
+
+	err := m.doStep(ctx, m.stepPosition < m.targetStepPosition, m.rotationPerMinute)
+	if err != nil {
+		return time.Second, fmt.Errorf("error stepping %w", err)
+	}
+	k := time.Duration(int64(time.Microsecond*1000*1000) / int64(math.Abs(float64(m.targetStepsPerSecond))))
+	return k, nil
+}
+
+// doStep has to be locked to call.
+// Depending on the direction, doStep will either treverse the stepSequence array in ascending
+// or descending order.
+func (m *uln2003) doStep(ctx context.Context, forward bool, rpm float64) error {
+	if forward {
+		for tick := 0; tick < len(stepSequence); tick++ {
+			err1 := m.in1.Set(ctx, stepSequence[tick][0], nil)
+			if err1 != nil {
+				return errors.Errorf("failed to set In1 with error in motor (%s)", m.motorName)
+			}
+
+			time.Sleep(minDelayBetweenTicks)
+			err2 := m.in2.Set(ctx, stepSequence[tick][1], nil)
+			if err2 != nil {
+				return errors.Errorf("failed to set In2 with error in motor (%s)", m.motorName)
+			}
+
+			time.Sleep(minDelayBetweenTicks)
+
+			err3 := m.in3.Set(ctx, stepSequence[tick][2], nil)
+			if err3 != nil {
+				return errors.Errorf("failed to set In3 with error in motor (%s)", m.motorName)
+			}
+
+			time.Sleep(minDelayBetweenTicks)
+
+			err4 := m.in4.Set(ctx, stepSequence[tick][3], nil)
+			if err4 != nil {
+				return errors.Errorf("failed to set In4 with error in motor (%s)", m.motorName)
+			}
+		}
+		time.Sleep(time.Duration(m.setStepperDelay(rpm)))
+		m.stepPosition++
+	} else {
+		for tick := len(stepSequence) - 1; tick >= 0; tick-- {
+			err1 := m.in1.Set(ctx, stepSequence[tick][0], nil)
+			if err1 != nil {
+				return errors.Errorf("failed to set In1 with error in motor (%s)", m.motorName)
+			}
+
+			time.Sleep(minDelayBetweenTicks)
+
+			err2 := m.in2.Set(ctx, stepSequence[tick][1], nil)
+			if err2 != nil {
+				return errors.Errorf("failed to set In2 with error in motor (%s)", m.motorName)
+			}
+
+			time.Sleep(minDelayBetweenTicks)
+
+			err3 := m.in3.Set(ctx, stepSequence[tick][2], nil)
+			if err3 != nil {
+				return errors.Errorf("failed to set In3 with error in motor (%s)", m.motorName)
+			}
+
+			time.Sleep(minDelayBetweenTicks)
+
+			err4 := m.in4.Set(ctx, stepSequence[tick][3], nil)
+			if err4 != nil {
+				return errors.Errorf("failed to set In4 with error in motor (%s)", m.motorName)
+			}
+		}
+		time.Sleep(time.Duration(m.setStepperDelay(rpm)))
+		m.stepPosition--
+	}
+
+	return nil
+}
+
+// GoFor instructs the motor to go in a specific direction for a specific amount of
+// revolutions at a given speed in revolutions per minute. Both the RPM and the revolutions
+// can be assigned negative values to move in a backwards direction. Note: if both are negative
+// the motor will spin in the forward direction.
+func (m *uln2003) GoFor(ctx context.Context, rpm, revolutions float64, extra map[string]interface{}) error {
```

Comment:
[major] can we clean this up a lot? We don't need the internal private functions to match what `gpiostepper` has. `GoFor` can just run the `doCycle` with a for loop and update the motor struct with the current position.

You don't need to have `goForInternal` at all I think.

Reply:
We don't want to overload GoFor function. goForInternal is a good way to keep the calculations for targetStepPosition and stepperDelay separated.  So I will vote for keeping goForInternal.